### PR TITLE
List tabs

### PR DESCRIPTION
**Pull request by kyori19**
image
image
image
image

Add ability to add list timelines to tabs.

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/db/Converters.kt [hunk 1]
```diff
@@ -61,13 +63,13 @@ class Converters {
...
                     val data = it.split(":")
-                    createTabDataFromId(data[0], data.drop(1))
+                    createTabDataFromId(data[0], data.drop(1).map { s -> URLDecoder.decode(s, "UTF-8") })
                 }
     }
 
     @TypeConverter
     fun tabDataToString(tabData: List<TabData>?): String? {
-        return tabData?.joinToString(";") { it.id + ":" + it.arguments.joinToString(":") }
+        return tabData?.joinToString(";") { it.id + ":" + it.arguments.joinToString(":") { s -> URLEncoder.encode(s, "UTF-8") } }
```
Comment (kyori19) [inline app/src/main/java/com/keylesspalace/tusky/db/Converters.kt]: List name may includes ":".

File: app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt [hunk 2]
```diff
@@ -200,6 +210,32 @@ class TabPreferenceActivity : BaseActivity(), Injectable, ItemInteractionListene
...
+        val adapter = ListSelectionAdapter(this)
+        mastodonApi.getLists()
+                .observeOn(AndroidSchedulers.mainThread())
+                .autoDispose(from(this, Lifecycle.Event.ON_DESTROY))
+                .subscribe (
+                        { lists ->
+                            adapter.addAll(lists)
+                        },
+                        {
+                        }
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt]: log an error here please

File: app/src/main/java/com/keylesspalace/tusky/db/Converters.kt (hunk 1 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/db/Converters.kt]: put that into an comment in the code please

File: app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+            view = layoutInflater.inflate(R.layout.item_autocomplete_hashtag, parent, false)
+        }
+        view!!
+
+        val list = getItem(position)
+        if (list != null) {
+            val title = view.hashtag
+            title.text = list.title
+            val icon = context.getDrawable(R.drawable.ic_list)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt]: The icon should have the same color as the text. You can use `ThemeUtils.getTintedDrawable`, or `IconicsDrawable`, see here
https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/ListsActivity.kt#L237-L238

File: app/src/main/java/com/keylesspalace/tusky/db/Converters.kt [hunk 4]
```diff
@@ -61,13 +63,13 @@ class Converters {
         return str?.split(";")
                 ?.map {
                     val data = it.split(":")
-                    createTabDataFromId(data[0], data.drop(1))
+                    createTabDataFromId(data[0], data.drop(1).map { s -> URLDecoder.decode(s, "UTF-8") })
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/db/Converters.kt]: Can you try moving the url decoding somewhere else, so it doesn't run for tabs that don't need it? E.g. to `TabData.createTabDataFromId`

File: app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import com.keylesspalace.tusky.entity.MastoList
+import kotlinx.android.synthetic.main.item_autocomplete_hashtag.view.*
+
+class ListSelectionAdapter(context: Context) : ArrayAdapter<MastoList>(context, R.layout.item_autocomplete_hashtag) {
+    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+        var view = convertView
+
+        if (convertView == null) {
+            val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+            view = layoutInflater.inflate(R.layout.item_autocomplete_hashtag, parent, false)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt]: I think its better to use a dedicated layout here. No bold text, more spacing between the CompoundDrawable and the text.

**Comment by kyori19** [comment]
@connyduck All done!

File: app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+            view = layoutInflater.inflate(R.layout.item_picker_list, parent, false)
+        }
+        view!!
+
+        val list = getItem(position)
+        if (list != null) {
+            val title = view.title
+            title.text = list.title
+            val icon = ThemeUtils.getTintedDrawable(context, R.drawable.ic_list, android.R.attr.textColorPrimary)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/adapter/ListSelectionAdapter.kt]:
```suggestion
            val icon = ThemeUtils.getTintedDrawable(context, R.drawable.ic_list, android.R.attr.textColorTertiary)
```

File: app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt [hunk 7]
```diff
@@ -200,6 +210,33 @@ class TabPreferenceActivity : BaseActivity(), Injectable, ItemInteractionListene
...
+        val adapter = ListSelectionAdapter(this)
+        mastodonApi.getLists()
+                .observeOn(AndroidSchedulers.mainThread())
+                .autoDispose(from(this, Lifecycle.Event.ON_DESTROY))
+                .subscribe (
+                        { lists ->
+                            adapter.addAll(lists)
+                        },
+                        { throwable ->
+                            throwable.printStackTrace()
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/TabPreferenceActivity.kt]:
```suggestion
                            Log.e("TabPreferenceActivity", "failed to load lists", throwable)
```

File: app/src/main/java/com/keylesspalace/tusky/db/Converters.kt [hunk 8]
```diff
@@ -67,7 +68,8 @@ class Converters {
 
     @TypeConverter
     fun tabDataToString(tabData: List<TabData>?): String? {
-        return tabData?.joinToString(";") { it.id + ":" + it.arguments.joinToString(":") }
+        // List name may include ":"
+        return tabData?.joinToString(";") { it.id + ":" + it.arguments.joinToString(":") { s -> URLEncoder.encode(s, "UTF-8") } }
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/db/Converters.kt]: nah, please remove the url encoding from here, it will run for all tabs.

**Comment by kyori19** [comment]
@charlag Thanks!

@connyduck 
I moved the decoder, but it's a bit ugly.
Do you have any other good ideas?

**Comment by connyduck** [comment]
Hmm this sucks 🤔 
You could make `TabData` an open class and override the `arguments` for lists, but thats also quite ugly. And there is probably no other delimiter we could use, so its propably best to prefer simpler code over performance. Can you move the url en/decoding back into the converter? Sorry for the back&forth

**Comment by kyori19** [comment]
@connyduck 
Reverted all changes.

**Pull request merged**